slam-libraries/viam-cartographer/src/slam_service/slam_service.cc

```diff
@@ -420,7 +432,74 @@ std::string SLAMServiceImpl::GetNextDataFile() {
     return GetNextDataFileOnline();
 }
 
-void SLAMServiceImpl::RunSLAM() {
+void SLAMServiceImpl::StartSaveMap() {
+    if (map_rate_sec == std::chrono::seconds(0)) {
+        return;
+    }
+    thread_save_map_with_timestamp =
+        new std::thread([&]() { this->SaveMapWithTimestamp(); });
+}
+
+void SLAMServiceImpl::StopSaveMap() {
+    if (map_rate_sec == std::chrono::seconds(0)) {
+        return;
+    }
+    thread_save_map_with_timestamp->join();
+}
+
+void SLAMServiceImpl::SaveMapWithTimestamp() {
+    auto check_for_shutdown_interval_usec =
+        std::chrono::microseconds(checkForShutdownIntervalMicroseconds);
+    while (b_continue_session) {
+        auto start = std::chrono::high_resolution_clock::now();
+        const std::string filename_with_timestamp =
+            viam::io::MakeFilenameWithTimestamp(path_to_map);
+
+        if (offlineFlag && finished_processing_offline) {
+            {
+                std::lock_guard<std::mutex> lk(map_builder_mutex);
+                // Save the map in a pbstream file
+                bool ok = map_builder.map_builder_->SerializeStateToFile(
```

Comment:
[opt] since this is duplicate code should we try to make it a function? just a void function to save the map should be fine, and leave the !ok check inside of it.